### PR TITLE
Refactor DrivePolicy architecture and configuration

### PR DESCRIPTION
**Pull request by vcharraut**
- Updated DrivePolicy to use shared network architecture instead of split network.
- Changed input size parameters to specific sizes for ego, partner, lane, boundary, traffic control, and conditioning inputs.
- Modified encoder configuration to include activation functions and layer normalization options.
- Removed gigaflow architecture in favor of a more flexible encoder design.
- Adjusted observation size calculations to include counts of various features.
- Updated environment bindings and configuration files to reflect new parameter names and structures.
- Enhanced the DriveBackbone class to support new encoder configurations and pooling mechanisms.
- Updated the Drive class to accommodate changes in the backbone initialization and observation encoding.

### CODE REVIEW

File: pufferlib/config/ocean/drive.ini [hunk 1]
```diff
@@ -11,21 +11,31 @@ num_workers = auto
...
+target_input_size = 64
+; Encoder activation - options: "relu", "tanh", "gelu"
+encoder_activation = "relu"
+encoder_layer_norm = True
+mask_padded_features = False
 ; Shared backbone layer
 backbone_hidden_size = 512
 backbone_num_layers = 4
+; Backbone activation - options: "relu", "tanh", "gelu"
+backbone_activation = "gelu"
```
Comment (eugenevinitsky) [inline pufferlib/config/ocean/drive.ini:29]: was gelu the winner?

Comment (vcharraut) [reply pufferlib/config/ocean/drive.ini:29]: I haven't test it enough tbf

File: tests/smoke_tests/test_drive_train.py [hunk 2]
```diff
@@ -148,7 +148,12 @@ def _build_config():
     _set_existing(
         args["policy"],
         {
-            "input_size": 32,
+            "ego_input_size": 32,
+            "partner_input_size": 32,
+            "lane_input_size": 32,
+            "boundary_input_size": 32,
+            "traffic_control_input_size": 32,
+            "target_input_size": 32,
```
Comment (eugenevinitsky) [inline tests/smoke_tests/test_drive_train.py]: suggestion; make these different numbers. Errors can become silent when everything has the same shape

File: pufferlib/ocean/torch.py [hunk 3]
```diff
@@ -19,34 +21,55 @@ class DriveBackbone(nn.Module):
...
+            return encoder(objects).max(dim=1).values
+
+        valid_mask = torch.arange(objects.shape[1], device=objects.device) < valid_counts.unsqueeze(1)
+        encoded_objects = objects.new_full(
+            (objects.shape[0], objects.shape[1], out_size),
+            torch.finfo(objects.dtype).min,
+        )
+        encoded_objects[valid_mask] = encoder(objects[valid_mask])
+        pooled = encoded_objects.amax(dim=1)
+        return torch.where(valid_counts.unsqueeze(1) == 0, encoded_objects.new_zeros(()), pooled)
```
Comment (eugenevinitsky) [inline pufferlib/ocean/torch.py:43]: this might deserve a unit test; there's some indexing going on here that makes me nervous

File: pufferlib/ocean/torch.py [hunk 4]
```diff
@@ -64,55 +87,47 @@ def __init__(
...
-                input_size,
-                encoder_gigaflow,
+                self.traffic_control_features_after_onehot, traffic_control_input_size
             )
-            num_feature_sets += 1
-        if self.conditioning_dim > 0:
-            self.conditioning_encoder = self._create_encoder(self.conditioning_dim, input_size, encoder_gigaflow)
-            num_feature_sets += 1
+            encoders_out += traffic_control_input_size
+        if self.target_dim > 0:
```
Comment (eugenevinitsky) [inline pufferlib/ocean/torch.py]: this name feels more confusing than conditioning?

File: pufferlib/ocean/torch.py [hunk 5]
```diff
@@ -64,55 +87,47 @@ def __init__(
...
-                self.traffic_control_features_after_onehot,
-                input_size,
-                encoder_gigaflow,
+                self.traffic_control_features_after_onehot, traffic_control_input_size
             )
-            num_feature_sets += 1
-        if self.conditioning_dim > 0:
-            self.conditioning_encoder = self._create_encoder(self.conditioning_dim, input_size, encoder_gigaflow)
-            num_feature_sets += 1
+            encoders_out += traffic_control_input_size
```
Comment (eugenevinitsky) [inline pufferlib/ocean/torch.py:112]: err, I mean the variable name target_dim is unclear. What's a target?

File: pufferlib/ocean/torch.py (hunk 4 above)
Comment (vcharraut) [reply pufferlib/ocean/torch.py]: idk, this encoder is supposed to have the goal + conditioning

in my mind it means it is a target for the agent to match, like he should get to the goal and he should act to match the coefficients

but if it is not clear enough i'll revert, the naming is very confusing

File: pufferlib/config/ocean/drive.ini [hunk 6]
```diff
@@ -111,7 +121,7 @@ reward_lane_align = 0.025
 reward_vel_align = 1.0
 reward_lane_center = 0.0038
 reward_center_bias = 0.0
-reward_velocity = 0.1
+reward_velocity = 0.0025
```
Comment (riccardosavorgnan) [inline pufferlib/config/ocean/drive.ini:124]: This seems like an important change, what's the justification? I agree it might have been way to high but I wonder if you tested anything and whether it's worth leaving a comment on it.

Comment (vcharraut) [reply pufferlib/config/ocean/drive.ini:124]: the old value is actually a mistake
the default coefficients should match the GIGAFLOW evaluation's

File: pufferlib/ocean/drive/drive.h [hunk 7]
```diff
@@ -4723,10 +4702,9 @@ static int write_partner_obs(Drive *env, Agent *ego, int agent_idx, float *obs,
 
     for (int j = 0; j < partners_to_write; j++) {
         Agent *other = &env->agents[nearby_agents[j].index];
-        float rel_x, rel_y, rel_heading_x, rel_heading_y, rel_vx, rel_vy;
```
Comment (riccardosavorgnan) [inline pufferlib/ocean/drive/drive.h:4726]: Are we removing the relative partner velocity from the observation? 
If so, it's probably worth documenting why since we always had it?

Comment (vcharraut) [reply pufferlib/ocean/drive/drive.h:4726]: the partner velocity was not used before this PR, we use sim_signed_speed instead

File: pufferlib/ocean/torch.py [hunk 8]
```diff
@@ -19,34 +21,55 @@ class DriveBackbone(nn.Module):
...
+    def _create_encoder(self, in_features, out_size):
+        layers = [pufferlib.pytorch.layer_init(nn.Linear(in_features, out_size))]
+        if self.encoder_layer_norm:
+            layers.append(nn.LayerNorm(out_size))
+        layers.append(self.encoder_act_cls())
+        layers.append(pufferlib.pytorch.layer_init(nn.Linear(out_size, out_size)))
+        return nn.Sequential(*layers)
+
+    def _encode_and_pool(self, objects, valid_counts, encoder, out_size):
+        if not self.mask_padded_features:
```
Comment (riccardosavorgnan) [inline pufferlib/ocean/torch.py:33]: probably worth activating mask padded features in the default config?

Comment (vcharraut) [reply pufferlib/ocean/torch.py:33]: The mask padded features option doesn't really improve perfomance and is not compatible with torch.compile for now. 

I added it because it remains a base code for anyone that want to improve it

**Review by eugenevinitsky — APPROVED** [review]
